### DISCUSSION TITLE
Dumping an image of Clog + clog:run-demo 1

### DISCUSSION THREAD

**Discussion by aykaramba** [rabbibotton/clog#74] [Q&A] [answered]
HI.

Clog comes with a really nice seat of tutorials.  I would like to learn how to dump an image of clog and have the image automatically run demo 1.  I want to learn how an executable binary could be created and run.

I have tried:

1) SBCL - started sbcl + (ql:quickload :clog) + (clog:run-demo 1)

When I do this, the sbcl command prompt is utterly controlled by the demo and I cannot get another command prompt to evaluate the dump image form.

Here is the output for reference:

* (clog:run-demo 1)
 ---- The demo src is located at: /home/.../clog-20210807-git/./demos/01-demo.lisp
Hunchentoot server is started.
Listening on 0.0.0.0:8080.
HTTP listening on    : 0.0.0.0:8080
HTML Root            : /home/.../clog-20210807-git/./static-files/
Boot js source       : compiled in
Boot file for path / : /boot.html
NIL
NIL
0
* New connection id - 1 - #<SERVER {10044C4373}>

sbcl is stuck at the <SERVER> part, no new command prompt ever appears so I can dump theimage.

2) In emacs i startup slibe + (ql:quickload :clog) + (clog:run-demo 1)

Accoding to google, you cannot dump an image from slime, you are supposed to switch to the *inferior-lisp* buffer and evaluate the dump image form there to dump the image.

However, here to, I have no access to a command prompt of any kind.  Instead the entire buffer is completely owned by some kind of signal from sbcl or slime or whatever this might be.

Here is what it looks like:

; compilation unit finished
;   Undefined function:
;     POSITIONAL-ARGS-NUMBER
;   caught 1 STYLE-WARNING condition

; file: /usr/share/common-lisp/source/slime/contrib/swank-arglists.lisp
; in: DEFUN PROVIDED-ARGUMENTS-REF
;     (SWANK::POSITIONAL-ARGS-NUMBER SWANK/BACKEND:ARGLIST)
;
; caught STYLE-WARNING:
;   undefined function: SWANK::POSITIONAL-ARGS-NUMBER
;
; compilation unit finished
;   Undefined function:
;     POSITIONAL-ARGS-NUMBER
;   caught 1 STYLE-WARNING condition

; file: /usr/share/common-lisp/source/slime/contrib/swank-arglists.lisp
; in: DEFUN PARSE-RAW-FORM
;     (SWANK::READ-CONVERSATIVELY SWANK::ELEMENT)
;
; caught STYLE-WARNING:
;   undefined function: SWANK::READ-CONVERSATIVELY
;
; compilation unit finished
;   Undefined function:
;     READ-CONVERSATIVELY
;   caught 1 STYLE-WARNING condition

Does anyone know / have instructions on how to dump a clog image and have it run a specific demo / app?

I am running Debian 11 + default emacs + default slime from the Debian repos.

NOTE: My apologies if this post is a dupe of an earlier post about the same issue from a few minutes ago.  I posted this question earlier and it either did not appear or dissapeard, so I am posting once again to see how the discussion forums on github work.

Thanks.

**Comment by smvamse** [comment]
Does "dumping an image" mean compiling?

**Comment by rabbibotton** [comment]
Code is actually compiled as loaded in most cases to CL environments. An
image is the entire environment packaged as an executable.

Yes it can be done. I will see if can add directions or a tutorial soon.

**Reply by aykaramba** [reply]
That would be awesome! Thank you very kindly for creating Clog.

**Comment by rabbibotton** [comment]
I haven't had time yet to produce a sample. I did take a moment to read your post more thoroughly. First to do a dump you have to do it from the command line not in emacs. Second here is an example using sbcl

(defun main ()
  (princ "Hello, world!"))

(sb-ext:save-lisp-and-die "hello.exe" :toplevel #'main :executable t)

**Reply by aykaramba** [reply]
Correct.  However, when you load up a CLOG example file, this isn't possible.   The SBCL comamand prompt is inaccessible as per point #1.  I get the Hunchentoot server start message as per description in point #1 and I cannot enter any commands at all.  I can only try and Ctrl-c out of that or close the terminal window.

No image can be dumped after a demo is started.

What am I not understanding?

NOTE: I am on Linux (Debian) with the standard SBCL from the repos.  I believe you develop on MacOS correct?  When you are in sbcl and you load up CLOG and run (clog:run-demo 1) does your SBCL command prompt provide feedback as per above and then open up a free prompt so that you can continue evaluating forms?

Mine doesn't do that.

I have reviewed a bunch of the options to dump an image and the other thing that can be done from SBCL is load up CLOG and dump the CLOG image and include a call to (clog:run-demo 1), but that image starts up and then crashes on a thread connection error.

If you like I can run as many test permutations necessary to explain this better.

Thank you kindly for your time.

**Reply by rabbibotton** [reply]
I should be able to look at this. Sorry for the delays

**Reply by aykaramba** [reply]
No need to apologize.  We are all grateful for the existence of CLOG.  Thank YOU kindly for taking interest in our questions.

**Reply by aykaramba** [reply]
To provide a more concrete example of running a CLOG application from sbcl but not receiving the release of the command prompt, here is what happens on Linux when I run the (clog-repl) application:

//--> start copy-paste

user@computer ~ $ sbcl
This is SBCL 2.1.1.debian, an implementation of ANSI Common Lisp.
More information about SBCL is available at <http://www.sbcl.org/>.

SBCL is free software, provided as is, with absolutely no warranty.
It is mostly in the public domain; some portions are provided under
BSD-style licenses.  See the CREDITS and COPYING files in the
distribution for more information.
* (ql:quickload :clog)
To load "clog":
  Load 1 ASDF system:
    clog
; Loading "clog"
.............
(:CLOG)
* (in-package clog-user)
#<PACKAGE "CLOG-USER">
* (clog-repl)
Hunchentoot server is started.
Listening on 0.0.0.0:8080.
HTTP listening on    : 0.0.0.0:8080
HTML Root            : /home/user/mnt/work/desktop/user/quicklisp/local-projects/clog/./static-files/
Boot js source       : compiled in
Boot file for path / : /debug.html
Use clog-user:*body* to access the clog-repl window.
NIL
* New connection id - 1 - #<SERVER {100400EB03}>

//--> end copy-paste

As you can see, in this example starting the (clog-repl) application starts up a browser session, it creates a new connection and a command prompt is not released so that forms can be typed in and evaluated.

I hope that helps.  I am working through the examples and trying to learn more to be able to assist.

**Reply by aykaramba** [reply]
One more datapoint:

1) In the previous notes, I described 3 ways that I cannot get access to the command line once a demo or a tutorial is running.

2) Here are the steps to how to dump an SBCL image once CLOG is quick loaded and what happens when I try to run the image:

a) Start sbcl: sbcl

b) Quickload CLOG: (ql:quickload :clog)

c) Dump the CLOG image and tell it to run (clog:run-demo 1) on execution:

(sb-ext:save-lisp-and-die 
  "clog"
  :compression t
  ;; this is the main function:
  :toplevel (lambda () 
              (clog:run-demo 1)                                 
              0)
  :executable t)

d) Run dumped CLOG image: ./clog

e)  Error message:

 ---- The demo src is located at: /home/user/mnt/work/desktop/user/quicklisp/local-projects/clog/./demos/01-demo.lisp
Hunchentoot server is started.
Listening on 0.0.0.0:8080.
HTTP listening on    : 0.0.0.0:8080
HTML Root            : /home/user/mnt/work/desktop/user/quicklisp/local-projects/clog/./static-files/
Boot js source       : compiled in
Boot file for path / : /boot.html
[2021-11-09 15:45:53 [ERROR]] Wake-for-shutdown connect failed: Condition USOCKET:CONNECTION-REFUSED-ERROR was signalled.
Unhandled SB-THREAD:JOIN-THREAD-ERROR in thread #<SB-THREAD:THREAD "clack-handler-hunchentoot" RUNNING
                                                   {10044C9AD3}>:
  Joining thread failed: thread #<THREAD "hunchentoot-listener-0.0.0.0:8080" ABORTED
                                   {1004F82F83}> did not return normally.

Backtrace for: #<SB-THREAD:THREAD "clack-handler-hunchentoot" RUNNING {10044C9AD3}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SB-THREAD:JOIN-THREAD-ERROR {1005885CB3}> #<unused argument> :QUIT NIL)
1: ((LAMBDA (SB-THREAD::C SB-THREAD::H) :IN SB-THREAD::%EXIT-OTHER-THREADS) #<SB-THREAD:JOIN-THREAD-ERROR {1005885CB3}> #<FUNCTION (LAMBDA (SB-THREAD::C SB-THREAD::H) :IN SB-THREAD::%EXIT-OTHER-THREADS) {527E044B}>)
2: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SB-THREAD:JOIN-THREAD-ERROR {1005885CB3}>)
3: (INVOKE-DEBUGGER #<SB-THREAD:JOIN-THREAD-ERROR {1005885CB3}>)
4: (ERROR SB-THREAD:JOIN-THREAD-ERROR :THREAD #<SB-THREAD:THREAD "hunchentoot-listener-0.0.0.0:8080" ABORTED {1004F82F83}> :PROBLEM :ABORT)
5: (SB-THREAD:JOIN-THREAD #<SB-THREAD:THREAD "hunchentoot-listener-0.0.0.0:8080" ABORTED {1004F82F83}> :DEFAULT NIL :TIMEOUT NIL)
6: ((:METHOD HUNCHENTOOT:SHUTDOWN (HUNCHENTOOT:ONE-THREAD-PER-CONNECTION-TASKMASTER)) #<HUNCHENTOOT:ONE-THREAD-PER-CONNECTION-TASKMASTER {10049A38C3}>) [fast-method]
7: ((:METHOD HUNCHENTOOT:STOP (HUNCHENTOOT:ACCEPTOR)) #<CLACK.HANDLER.HUNCHENTOOT::CLACK-ACCEPTOR (host 0.0.0.0, port 8080)> :SOFT NIL) [fast-method]
8: ((FLET "CLEANUP-FUN-3" :IN CLACK.HANDLER.HUNCHENTOOT:RUN)) [cleanup]
9: (CLACK.HANDLER.HUNCHENTOOT:RUN #<FUNCTION (LAMBDA (LACK.MIDDLEWARE.BACKTRACE::ENV) :IN "/home/user/mnt/work/desktop/user/quicklisp/dists/quicklisp/software/lack-20210807-git/src/middleware/backtrace.lisp") {1003A9AD5B}> :ALLOW-OTHER-KEYS T :ADDRESS "0.0.0.0" :PORT 8080 :DEBUG T :USE-THREAD T :ADDRESS "0.0.0.0")
10: ((LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS))
11: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
12: ((FLET "WITHOUT-INTERRUPTS-BODY-11" :IN SB-THREAD::RUN))
13: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
14: ((FLET "WITHOUT-INTERRUPTS-BODY-4" :IN SB-THREAD::RUN))
15: (SB-THREAD::RUN)
16: ("foreign function: call_into_lisp")
17: ("foreign function: funcall1")

unhandled condition in --disable-debugger mode, quitting

3) Conclusion: I finally figured out how to dump the image in a non-interactive / non-running demo.  Instead, just loading up CLOG dumping the image with the clog:run-demo 1 command was a workaround to see if we can get a running local binary.  Hopefully the error message and steps provide a way to replicate the process, perhaps I am doing something wrong.

**Reply by kaeland** [reply]
Thank you for providing these steps. I hope a solution can be found soon.

**Reply by smvamse** [reply]
Put this code in the `:toplevel` function.
Here `main` is the `:toplevel` function.
![Screenshot 2021-11-11 at 2 52 03 PM](https://user-images.githubusercontent.com/84125167/141271882-1f9fdcee-a2d7-4d30-b122-bce19d6424b9.png)

The `handler-case` part will fix the error.

**Comment by aykaramba** [comment, marked as answer]
Okay here is the working solution for dumping a clog image to an executable for total noobs:

1) Start sbcl:  user@computer >> sbcl

2) Quickload clog:  (ql:quickload :clog)

3) Copy / paste this code to your sbcl command prompt:

```
(sb-ext:save-lisp-and-die
 "clog"
 :compression t
 ;; this is the main function:
 :toplevel (lambda ()
	     ;; tell clog to run one of the demos
	     ;; you can use (clog:run-demo 1) or (clog:run-demo 2), etc.
	     (clog:run-demo 3)

	     ;; this part handles the threading errors
	     (handler-case (bt:join-thread (find-if (lambda (th)
						      (search "hunchentoot" (bt:thread-name th)))
						    (bt:all-threads)))
	       ;; Catch a user's C-c
	       (#+sbcl sb-sys:interactive-interrupt
		#+ccl  ccl:interrupt-signal-condition
		#+clisp system::simple-interrupt-condition
		#+ecl ext:interactive-interrupt
		#+allegro excl:interrupt-signal
		() (progn
		     (format *error-output* "Aborting.~&")
		     (clack:stop *server*)
		     (uiop:quit)))
	       (error (c) (format t "Woops an unknown error occured" c)))

	     0)
 :executable t)
```

4) Notes:

a) You will get an executable called "clog".  Just change the "clog" part above to whatever you want to name your executable

b) You can run whatever demo  you want with this line: (clog:run-demo 3)  Just read the README here and some of the manuals to figure out how to run some of the demos and tutorials.

c) Apparently, with a common lisp image, we can start up the application and then connect to it using SWANK and fiddle with the running binary.   I haven't tried that yet, I guess that will be the next dumb thread that I start here. 

:)


5) The most amazing thing about CLOG and being able to compile it to an executable binary is that if you code up everything in CLOG you have only a single binary to run and you don't have to worry about includes and paths for js and html for the running code.  So nice.

If we do decide to use js / html / media resources though, it should be no problem either, you can have all of them in a folder structure and use relative paths for the binary to access.

Lastly, I have to try and compile CLOG now on an arm device of some sort.  SBCL already runs sooooo ... this should be fun!

I just love having a web server and framework all in one binary.   This is so nice.

Thanks SMVAMSE, greatly appreciate that insight.

Thanks to Rabbi Botton for CLOG.

**Reply by kaeland** [reply]
Sweet!! Thank you for the writeup. I'll have to give this a try soon. :D

**Reply by smvamse** [reply]
> Thanks SMVAMSE, greatly appreciate that insight.

You're welcome @aykaramba!!

**Comment by aykaramba** [comment]
So, I just managed to dump a clog image on one of these:  https://shop.pocketchip.co/collections/frontpage/products/pocket-c-h-i-p-new

Details:

* Device is upgraded to Debian 11.  It's an old dead crowdsourced project.
* The cpu is a single core, 32 bit, kinda sad ARMv7 processor.
* SBCL could not compile things like cffi and ssl support.  

However, the image was dumped successfully and does actually execute and stay running.  None of the demos / tutorials run because (I think) the key issue is threading support in SBCL on the sbcl version available in the ARMHF release of Debian.

So, while it doesn't actually render out pages successfully, the code compiles and an image can successfully be dumped and started up.

Why?

Because.

:)

**Reply by rabbibotton** [reply]
I frequently use clog in executables, but I add to depends-on in asdf "clack-handler-hunchentoot"
